litecommands-core/src/dev/rollczi/litecommands/scheduler/SchedulerExecutorPoolImpl.java

```diff
@@ -13,79 +18,137 @@
 
 public class SchedulerExecutorPoolImpl implements Scheduler {
 
-    private static final String MAIN_THREAD_NAME_FORMAT = "scheduler-%s-main";
-    private static final String ASYNC_THREAD_NAME_FORMAT = "scheduler-%s-async-%d";
+    private final ScheduledExecutorService mainExecutor;
+    private final boolean closeMainExecutorOnShutdown;
 
-    private final ThreadLocal<Boolean> isMainThread = ThreadLocal.withInitial(() -> false);
-
-    private final ExecutorService mainExecutor;
-    private final ExecutorService asyncExecutor;
-    private final ScheduledExecutorService scheduledExecutor = Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors());
+    private final ScheduledExecutorService asyncExecutor;
+    private final boolean closeAsyncExecutorOnShutdown;
 
+    @Deprecated
     public SchedulerExecutorPoolImpl(String name) {
         this(name, -1);
     }
 
+    @Deprecated
     public SchedulerExecutorPoolImpl(String name, int pool) {
         this.mainExecutor = Executors.newSingleThreadScheduledExecutor(runnable -> {
             Thread thread = new Thread(runnable);
-            thread.setName(String.format(MAIN_THREAD_NAME_FORMAT, name));
-
+            thread.setName(String.format("scheduler-%s-main", name));
             return thread;
         });
-
-        this.mainExecutor.submit(() -> isMainThread.set(true));
+        this.closeMainExecutorOnShutdown = true;
 
         AtomicInteger asyncCount = new AtomicInteger();
         ThreadFactory factory = runnable -> {
             Thread thread = new Thread(runnable);
-            thread.setName(String.format(ASYNC_THREAD_NAME_FORMAT, name, asyncCount.getAndIncrement()));
+            thread.setName(String.format("scheduler-%s-async-%d", name, asyncCount.getAndIncrement()));
 
             return thread;
         };
 
-        this.asyncExecutor = pool < 0 ? Executors.newCachedThreadPool(factory) : Executors.newFixedThreadPool(pool, factory);
+        this.asyncExecutor = Executors.newScheduledThreadPool(
+            Math.max(pool, Runtime.getRuntime().availableProcessors()) / 2,
+            factory
+        );
+        this.closeAsyncExecutorOnShutdown = true;
```

Comment:
Do not change the way this class works and do not adapt its operation for unit tests. If you want you can create a new implementation but do not violate the current API and its implementation.

Your implementation splits 2 pools, both of which have multiple threads, it has nothing to do with the "main" thread and can lead to concurrency in unexpected places.

Reply:
Let's do it this way, let's keep `SchedulerExecutorPoolImpl` in the form it was. Let's add a separate implementation: `SchedulerFullAsyncImpl` :P

Reply:
ok sure